checks.d/disk.py

```diff
@@ -0,0 +1,229 @@
+        self._load_legacy_option(instance, 'excluded_disk_re', '^$',
+                                 legacy_name='device_blacklist_re',
+                                 operation=re.compile)
+
+    def _load_legacy_option(self, instance, option, default,
+                            legacy_name=None, operation=lambda l: l):
+        self_option = instance.get(option, default)
+        if legacy_name is None:
+            legacy_name = option
+        legacy_option = self.agentConfig.get(legacy_name)
+        if self_option == default and legacy_option is not None:
+            self_option = legacy_option
+        setattr(self, '_{0}'.format(option), operation(self_option))
+
+    def collect_metrics_psutil(self):
+        self._valid_disks = {}
+        for part in psutil.disk_partitions(all=self._all_partitions):
+            # we check all exclude conditions
+            if self._exclude_disk_psutil(part):
+                continue
+            else:
+                # For later, latency metrics
+                self._valid_disks[part.device] = [part.fstype, part.mountpoint]
+            self.log.debug('Passed: {0}'.format(part.device))
+
+            tags = [part.fstype] if self._tag_by_filesystem else []
+            device_name = part.mountpoint if self._use_mount else part.device
+            for metric_name, metric_value in self._collect_part_metrics(part):
+                self.gauge(metric_name, metric_value,
+                           tags=tags, device_name=device_name)
+        # And finally, latency metrics, a legacy gift from the old Windows Check
+        self.collect_latency_metrics()
+
+    def _exclude_disk_psutil(self, part):
+        # skip cd-rom drives with no disk in it; they may raise
+        # ENOENT, pop-up a Windows GUI error for a non-ready
+        # partition or just hang;
+        # and all the other excluded disks
+        return ((Platform.is_win32() and ('cdrom' in part.opts or
+                                          part.fstype == '')) or
+                self._exclude_disk(part.device, part.fstype))
+
+    # We don't want all those incorrect devices
+    def _exclude_disk(self, name, filesystem):
+        return (name in self.FAKE_DEVICES or
+                name in self._excluded_disks or
+                self._excluded_disk_re.match(name) or
+                filesystem in self._excluded_filesystems)
+
+    def _collect_part_metrics(self, part):
+        usage = psutil.disk_usage(part.mountpoint)
+        metrics = {}
+        for name in ['total', 'used', 'free']:
+            # For legacy reasons,  the standard unit it kB
+            metrics[self.METRIC_DISK.format(name)] = getattr(usage, name) / 1024.0
+        # FIXME: 6.x, use percent, a lot more logical than in_use
+        metrics[self.METRIC_DISK.format('in_use')] = usage.percent / 100.0
+        if Platform.is_unix():
+            metrics.update(self._collect_inodes_metrics(part.mountpoint))
+
+        return metrics.iteritems()
```

Comment:
I am not sure it's adding much here to return a generator, you should return a dictionary and iterate on it in the caller.
Otherwise it's confusing some `_collect_*` methods return dicts, other return dictionaries or call `self.gauge` directly.